### ISSUE TITLE
Warm-starting moving averages (batch norm) as part of an estimator not possible

### ISSUE THREAD

**Issue by jonasrauber** [tensorflow/tensorflow#19903]
I want to use a `tf.estimators.Estimator` to fine-tune a model that contains batch normalization (e.g. ResNet). To initialize the model, I use the new `WarmStartSettings` and pass it to the estimator's `warm_start_from` argument. Unfortunately, this will only warm-start trainable variables, and the `moving_mean` and `moving_variance` created by the batch normalization layer are not part of the trainable variables collection. Thus, the moving averages will not be warm-started.

Because of this problem, TensorFlow 1.9 (rc0) introduced the possibility to pass a list of variables to the `WarmStartSettings`. Unfortunately, when using estimators, this does not help because the variables are recreated all the time and not know at the position where the warm start settings have to be defined.

A possible solution might be to make it possible to pass a function to the `warm_start_from` argument of estimators that has access to the current graph and returns a `WarmStartSettings` object.

My current workaround is to specify a list of variable names rather than variables in the `WarmStartSettings`, but this is not exactly how it's supposed to be and comes with its own problems (e.g. getting the list of variable names before the model was built; I just use all the variable names that are saved in the checkpoint and exlude e.g. `global_step`, but this is problematic because it circumvents certain checks and assertions).

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by jonasrauber** [comment]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: all up to 1.9rc0
- **Python version**: N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
See also https://github.com/keras-team/keras/pull/9965#issuecomment-398183440 /cc @ppwwyyxx

Check also reviews/comments in https://openreview.net/forum?id=Hk6dkJQFx

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by eddie-zhou** [comment]
Assuming you are writing your own `model_fn` (not using a canned `tf.Estimator`), can you try not using `tf.estimator.WarmStartSettings` during estimator init, and instead directly using `tf.train.warm_start` inside your `model_fn`?  That way you have direct access to the `variables` you've created inside the `model_fn`.  This is similar to directly using `tf.init_from_checkpoint`.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ispirmustafa, @eddie-zhou: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jiarenyf** [comment]
@eddie-zhou `tf.train.warm_start` excludes variables such as accumulators and moving statistics from batch norm.

I wonder why ?

And thus `tf.train.warm_start` is useless for this issues.

**Comment by eddie-zhou** [comment]
You're correct that it only includes vars in TRAINABLE_VARIABLES by default, but instead of the default regex, you can specify exactly what variables to warm-start, and include the accumulators and moving stats explicitly.

**Comment by bhack** [comment]
@eddie-zhou Honestly it is quite  a regular task so probably is better to have a new parameter

**Comment by eddie-zhou** [comment]
@bhack Could you make explicit your suggested API change?  Note that warm-starting accumulators is definitely inadvisable in many settings (given that new vocabulary entries have newly initialized accumulators), which is why we made the choice for the default to not include them.

**Comment by bhack** [comment]
I meant:
> you can specify exactly what variables to warm-start, and include the accumulators and moving stats explicitly

So I think that you can have a new parameter to recover accumulators and moving stats explicitly for the batch norm case without going to cherry picking variable with regex.

**Comment by jonasrauber** [comment]
I agree with @bhack. For many common real-world tasks related to warm-starting (fine-tuning existing networks, pretraining on a different dataset, etc.), it is important to restore not just the variables trained with backprop (i.e. the `trainable_variables`) but also the variables _trained_ using dataset statistics (e.g. accumulators, moving averages, etc.). Forgetting to restore them is such a common source of problems that are hard to notice; ideally it should be a very simple to switch between restoring `trainable_variables` and all _trainable variables_ (i.e. including accumulators, …) (as you notice: I even think it's a misnomer to only refer to some of them as _trainable_ variables).

**Comment by eddie-zhou** [comment]
Thanks for your thoughts, @jonasrauber!  I do agree that it can be useful in many settings, but by not being prescriptive in the API, we intentionally force users to explicitly think about exactly what variables they want to warm-start, and whether or not it is appropriate for their task.  Does that make sense?

**Comment by jonasrauber** [comment]
@eddie-zhou In principle I agree with the approach of _forcing explicit decisions_ (though I guess most ML researchers prefer more practical defaults like _restoring everything_). But in the case here, TensorFlow is not actually forcing an explicit decision, it's defaulting to a non-intuitive behavior (restoring layer-params, but not batch-norm params) that can cause silent bugs that may go unnoticed for a while (in fact, I have seen this happening to several people in our group). In practice, people will not notice that they haven't restored the batch norm params and assume they did.

**Comment by jonasrauber** [comment]
That _Note_ is certainly helpful, but maybe you can make it more prominent. Also: adding an explanation _how_ to change this (how to also restore accumulators) might help (in particular in situations like the one I described when opening this issue).

In any case: Thanks a lot for listening to our feedback :)

**Comment by bhack** [comment]
I think that frequent routines need to be automated in the API then flexibility could be useful for less frequent use cases.

**Comment by eddie-zhou** [comment]
FYI, an internal contribution has resulted in the following functionality being added:  vars_to_warm_start can now be "a function returning a boolean indicating whether a variable should warm-started. This function will be called for each item returned by `tf.get_collection(GLOBAL_VARIABLES)`".  This makes selecting non-trainable variables (like batch-norm) much easier.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @eddie-zhou: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @eddie-zhou: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @eddie-zhou: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ychfan** [comment]
Based on [current implementation](https://github.com/tensorflow/tensorflow/blob/f382bd6279141fe169e5aa8ab0867f18840ee037/tensorflow/python/training/warm_starting_util.py#L335), if a list of regex is passed as `vars_to_warm_start`, then all matched variables in GLOBAL_VARIABLES will be restored. I think the easiest way to restore everything is `vars_to_warm_start=['.*']`

**Comment by offbye** [comment]
I have tried to warm up variebles  resnet_v2_101  /BatchNorm/moving_mean  and   resnet_v2_101/block3/unit_13/bottleneck_v2/conv1/BatchNorm/moving_variance.

any good way to do it ?

**Comment by eddie-zhou** [comment]
Correction on my above statement: the internal contribution was not approved.

Re: @offbye -- can you provide the failure message?  Are you sure that those are your variable names?

**Comment by albanie** [comment]
Just flagging this here for others who hit the same issue.  The default behaviour affects warm starting from pre-trained Keras models that include batch norm (means and variances are not loaded after the warmstart) when calling [model_to_estimator()](https://www.tensorflow.org/api_docs/python/tf/keras/estimator/model_to_estimator). The function call happens [here](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/keras.py#L502-L504).

As a temporary fix, if you need to extract features with a pretrained model, you can patch around it with @ychfan’s suggestion - e.g. updating `tensorflow_estimator/python/estimator/keras.py` to use the code below, but for the general use-case this may be undesirable for other reasons:

```
  ws = estimator_lib.WarmStartSettings(
    ckpt_to_initialize_from=warm_start_path,
    vars_to_warm_start=['.*'],
  )
  estimator = estimator_lib.Estimator(keras_model_fn,
                                      config=config,
                                      warm_start_from=ws)
```

**Comment by Kublai-Jing** [comment]
Hi, @ychfan @albanie ,

 I am actually a bit confused here. Does the following code warm-starts ALL variables, including TRAINABLE and GLOBAL ones ? 
```
ws = estimator_lib.WarmStartSettings(
    ckpt_to_initialize_from=warm_start_path,
    vars_to_warm_start=['.*'],
  )
  estimator = estimator_lib.Estimator(keras_model_fn,
                                      config=config,
                                      warm_start_from=ws)

```
This doesn't seem to be the case in the code: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/warm_starting_util.py#L315

What

**Comment by muddham** [comment]
@jonasrauber Did you get chance to try the code suggested in previous comments. If it worked let us know  we will close this issue

**Comment by muddham** [comment]
@jonasrauber  Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by sharvil** [comment]
Since this issue was closed without an actual resolution, here's a workaround that warm starts all trainable variables and batch norm variables but none of the other global variables:

```
class WarmStartHook(tf.estimator.SessionRunHook):
  def __init__(self, checkpoint):
    self.checkpoint = checkpoint

  def begin(self):
    regex = '.*batch_normalization/(moving.*|gamma|beta):0$'
    tf.train.warm_start(self.checkpoint)
    tf.train.warm_start(self.checkpoint, [regex])
```

You can use this just like any other Estimator training hook.
